### ISSUE TITLE
Unexpected: validate = lambda val: {0,None,..} does not fail

### ISSUE THREAD

**Issue by mirko** [marshmallow-code/marshmallow#1458]
I wouldn't call this a bug but rather unexpected behaviour (for the time being).

When declaring a field and passing a (custom) validator, the validation only fails if the validator returns `False`.
This is to be expected, however I'd also expect the validation failing if the validator returns `0` or `None` or `[]` or `{}` or anything which usually evals to `False`.
For me it's common to `if` on such values usually evaluating to `False`.

Example:
`fields.Integer(required = True, validate=lambda val: 0)`
vs
`fields.Integer(required = True, validate=lambda val: bool(0))`

**Comment by lafrech** [comment]
After a quick search, I didn't find the docs for this.

From the existing validators in the code, it looks like a validator is supposed to either return the value if validation pass or raise ValidationError if validation fails. I don't understand why we return the value anyway.

What you're seeing is due to this piece of code:

```py
        for validator in self.validators:
            try:
                r = validator(value)
                if not isinstance(validator, Validator) and r is False:
                    raise self.make_error("validator_failed")
```

Apparently we also allow custom validator to just return `False` rather than a complete `ValidationError`, and the resulting error is a generic "Invalid value" message.

We could change the test to catch all falsy values

```py
                if not isinstance(validator, Validator) and not r:
```

but I'm not sure it would be a good idea. An empty list is falsy but can be considered valid. I like the explicitness of `False`.

**Comment by mirko** [comment]
Let me rephrase it then: *If* it is intended that way, I'd like to see it documented, as I still find it quite counter-intuitive. Happy to do so, but beforehand I'd like to ensure that it's actually intentional.

**Comment by lafrech** [comment]
I think it is intentional.

**Comment by sloria** [comment]
I improved the docs for this in https://github.com/marshmallow-code/marshmallow/pull/1774 .

I hesitated on this for a long time because I'm not sure the `False` behavior is actually a thing we should keep around. But ultimately, it doesn't hurt to document the current behavior.

I've opened a separate issue to remove the `False` behavior https://github.com/marshmallow-code/marshmallow/issues/1775
